include/node.tcc

```diff
@@ -272,7 +272,22 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_velocity_constraints() {
     const auto direction = static_cast<unsigned>(dir % Tdim);
     // Phase: Integer value of division (dir / Tdim)
     const auto phase = static_cast<unsigned>(dir / Tdim);
-    this->velocity_(direction, phase) = constraint.second;
-    this->acceleration_(direction, phase) = 0.;
+
+    if (!generic_velocity_constraints_) {
+      this->velocity_(direction, phase) = constraint.second;
+      this->acceleration_(direction, phase) = 0.;
+    } else {
+      // Transform to local coordinate
+      Eigen::Matrix<double, Tdim, Tnphases> local_velocity =
```

Comment:
`const Eigen::Matrix<double, Tdim, Tdim>  inverse_rotation_matrix_ = rotation_matrix_.inverse();`

Reply:
Is it a private variable? Better that way to avoid inverting the same small matrix over and over?

Reply:
Let's not add it to the private variables, because it's more memory. Please change here.